Mega cleanup for Cosmos DB

markjbrown opened the pull request:
## Description

This PR is a large clean up of all Cosmos DB PowerShell samples. This PR contains:
- Removal of PS samples not directly referenced in our docs.
- Superfluous samples not necessary for PS users.
- Fixes of existing samples.
- Random character generators for all samples which provision new resources.
- New samples that will be referenced in an upcoming PR in our docs.
- Moved samples that were in a different folder.

## Checklist

<!--
    Filling in this checklist is mandatory! If you don't, your pull request
    will be rejected without further review. Checklists must be completed
    within 7 days of PR submission.

    To check a box in markdown, make sure that it is formatted as [X] (no whitespace).
    Not formatting checkboxes correctly may break automated tools and delay PR processing.
-->

- [X] This pull request was tested on __both of__:
  - [X] PowerShell 5.1 (Windows)
  - [X] PowerShell 6.x ([Latest PowerShell](https://github.com/PowerShell/PowerShell/releases))
- [X] Scripts do not contain static passwords or other secret tokens.
- [X] All Azure resource identifiers which must be universally unique are guaranteed to be so.

### Testing information

<!--
    Each testing environment is a triplet:

        Platform, PowerShell version, Az version

    Copy/paste and fill in the following block for as many combinations of the above as you tested on.
-->

Platform:

Powershell version: `$PSVersionTable.PSVersion`

Az version: `(Get-InstalledModule -Name Az).Version`

markjbrown commented:
@sptramer  and @SnehaGunda  please review at your earliest convenience. This is the PR we discussed a few weeks ago to clean up all the Cosmos DB PS samples as well as implement the discussed changes to make them able to run without any user interaction.

Thank you both.

sptramer commented:
@SnehaGunda - Please @ me in a comment when you've finished reviewing from a technical perspective for CosmosDB and I'll make sure I prioritize handling this PR (since it was clearly a lot of work.) Thanks!

sptramer commented:
@markjbrown This PR may remain open for a while, since it's a good checkpoint for an automated test system. I'll be ignoring any style guide issues for cosmosdb PR submissions until this one is resolved as a result.

markjbrown commented:
ok, thanks.

Hey @SnehaGunda any ETA on when you'll be reviewing on your end?

Thanks.

SnehaGunda commented on cosmosdb/common/ps-account-firewall-create.ps1:
```diff
@@ -0,0 +1,34 @@
+# Create an Azure Cosmos Account for SQL (Core) API with IP Firewall
```
Not yet linked from doc page.

markjbrown replied:
Yes, cannot link until this is published.

SnehaGunda commented on cosmosdb/mongodb/ps-mongodb-ru-get.ps1:
```diff
@@ -1,13 +1,13 @@
 $databaseThroughputResourceName = $accountName + "/mongodb/" + $databaseName + "/throughput"
+$databaseThroughputResourceType = "Microsoft.DocumentDb/databaseAccounts/apis/databases/settings"
 $collectionThroughputResourceName = $accountName + "/mongodb/" + $databaseName + "/" + $collectionName + "/throughput"
+$collectionThroughputResourceType = "Microsoft.DocumentDb/databaseAccounts/apis/databases/collections/settings"
```
Is the $collectionThroughputResourceType variable not used in this script? you can delete if not used.

markjbrown replied:
Yes, it is used in the second Get operation here.

Get-AzResource -ResourceType $collectionThroughputResourceType `
    -ApiVersion $apiVersion -ResourceGroupName $resourceGroupName `
    -Name $collectionThroughputResourceName  | Select-Object Properties

SnehaGunda commented on cosmosdb/sql/ps-account-failover-priority-update.ps1:
```diff
@@ -1,14 +0,0 @@
-# Change the failover priority for an Azure Cosmos Account
```
This script is used by \azure-docs-pr\articles\cosmos-db\scripts\powershell\common\ps-account-failover-priority-update.md file

markjbrown replied:
Thanks for catching. I have created an updated version of this script. I will modify the samples page to point to the new one after this merges.

SnehaGunda commented on cosmosdb/sql/ps-account-firewall-create.ps1:
```diff
@@ -1,6 +1,6 @@
 # Create an Azure Cosmos Account for SQL (Core) API with IP Firewall
```
This script isn't used by any file in azure-docs-pr, can it be deleted? There is already another script "common/ps-account-firewall-create.ps1" with the same functionality I guess

markjbrown replied:
I will be adding a sample page for this once it is merged.

markjbrown replied:
Yes. This one can be deleted. I will be adding a new sample page to reference this once after this is merged.

SnehaGunda commented on cosmosdb/sql/ps-account-update.ps1:
```diff
@@ -1,4 +1,4 @@
-# Update an Azure Cosmos Account and add a region
```
There are two files in the azure-docs-pr repo that reference this file, we should remove one of them. 

\azure-docs-pr\articles\cosmos-db\scripts\powershell\sql\ps-account-update.md

\articles\cosmos-db\scripts\powershell\common\ps-account-update.md

markjbrown replied:
I will modify the path in the samples page once this PR is merged.

markjbrown commented on cosmosdb/cassandra/ps-cassandra-create.ps1:
```diff
@@ -1,12 +1,22 @@
-# Create a multi-master Azure Cosmos Account for Cassandra API with a keyspace with shared  
-# throughput, and a table with dedicated throughput with last writer wins conflict resolution policy
-$resourceGroupName = "myResourceGroup"
+# Create a zure Cosmos Account for Cassandra API with multi-master enabled, a keyspace with shared throughput,
```
Fixed in next commit

markjbrown commented on cosmosdb/cassandra/ps-cassandra-list-get.ps1:
```diff
@@ -1,34 +1,38 @@
-    -ApiVersion "2015-04-08" -ResourceGroupName $resourceGroupName `
+Get-AzResource -ResourceType $tableResourceType `
+    -ApiVersion $apiVersion -ResourceGroupName $resourceGroupName `
     -Name $tableResourceName | Select-Object Properties
```
Added a lf at the end of this file. Should correct this.

markjbrown commented on cosmosdb/cassandra/ps-cassandra-list-get.ps1:
```diff
@@ -1,34 +1,38 @@
 $tableName = "table1"
 $accountResourceName = $accountName + "/cassandra/"
 $keyspaceResourceName = $accountName + "/cassandra/" + $keyspaceName
+$keyspaceResourceType = "Microsoft.DocumentDb/databaseAccounts/apis/keyspaces"
```
It doesn't but it is consistent with the other examples.

markjbrown commented on cosmosdb/common/ps-account-add-region.ps1:
```diff
@@ -0,0 +1,49 @@
+# Create a Cosmos SQL API account in two regions, then add a third region
```
I need to change the name of this script to /common/ps-account-update. Then I will leave the current sample in powershell_scripts/cosmosdb/sql/ps-account-update.ps1 in place until after this and the docs update is merged. Then will delete that sample.

markjbrown replied:
Note: this rename will be in the next Commit.

markjbrown commented on cosmosdb/common/ps-account-failover-priority-update.ps1:
```diff
@@ -0,0 +1,44 @@
+# Change the failover priority for a single-master Azure Cosmos Account
```
Thanks for catching. Have updated the path in the Samples page to point to this one.

markjbrown replied:
Marking as unresolved. Will resolve this once this merges.

SnehaGunda replied:
I see you captured this change below in the summary.

markjbrown commented on cosmosdb/create-and-configure-Table-database/create-and-configure-Table-database.ps1:
```diff
@@ -1,44 +0,0 @@
-# Set the Azure resource group name and location
```
Can you point me to the article itself? I cannot find this.

SnehaGunda replied:
It's here: https://docs.microsoft.com/en-us/azure/cosmos-db//scripts/powershell/create-table-database-account-powershell?toc=%2fpowershell%2fmodule%2ftoc.json 

you can replace the file name in the URI to get to the right one.

SnehaGunda replied:
https://docs.microsoft.com/en-us/azure/cosmos-db//scripts/powershell/create-table-database-account-powershell?toc=%2fpowershell%2fmodule%2ftoc.json  - This page isn't linked from samples page as the script is deleted, we can delete the doc page. Adding this to your summary list below.  

https://docs.microsoft.com/en-us/azure/cosmos-db/scripts/powershell/table/ps-table-create?toc=%2fpowershell%2fmodule%2ftoc.json - This page is linked, so let's leave it as is.

markjbrown commented on cosmosdb/create-and-configure-cassandra-database/create-and-configure-cassandra-database.ps1:
```diff
@@ -1,44 +0,0 @@
-# Set the Azure resource group name and location
```
Cannot find this article referenced in published docs. I see the article itself but do not see how to navigate to it.

SnehaGunda replied:
The docs are published but looks like they aren't linked to from the navigator page. It's here:

https://docs.microsoft.com/en-us/azure/cosmos-db//scripts/powershell/create-and-configure-cassandra-database?toc=%2fpowershell%2fmodule%2ftoc.json

SnehaGunda replied:
Yes the docs are not linked but users can still find them with browser search because they are published. If we don't want them we should delete the .md files from the repo.

markjbrown commented on cosmosdb/create-and-configure-graph-database/create-and-configure-graph-database.ps1:
```diff
@@ -1,44 +0,0 @@
-# Set the Azure resource group name and location
```
Yes, same as the other similar ones. I think these can all be deleted.

SnehaGunda replied:
Published version is here: https://docs.microsoft.com/en-us/azure/cosmos-db//scripts/powershell/create-graph-database-account-powershell?toc=%2fpowershell%2fmodule%2ftoc.json

can you please check to see if it's safe to delete?

SnehaGunda replied:
yeah please delete them.

markjbrown commented on cosmosdb/create-and-configure-mongodb-database/create-and-configure-mongodb-database.ps1:
```diff
@@ -1,40 +0,0 @@
-# Set the Azure resource group name and location
```
Same comment as the others. If these are now orphaned I will delete.

SnehaGunda replied:
Published version is here: https://docs.microsoft.com/en-us/azure/cosmos-db//scripts/powershell/create-mongodb-database-account-powershell?toc=%2fpowershell%2fmodule%2ftoc.json

SnehaGunda commented:
@markjbrown is this PR ready for signoff?

markjbrown commented:
Hi Sneha. It seems the only remaining things was making sure the old PS scripts for Cassandra (2), Gremlin, MongoDB were deleted from docs repo. I have this on my To Do so I think we are ready to merge.

Please take one more pass at the unresolved comments and go ahead and do sign-off. Once this is merged. I will do the following on the docs side.

Create/update the following articles:
- New Samples page to point to /common/ps-account-firewall-create.ps1
- New Samples page to point to /common/ps-account-firewall-create.ps1
- Change existing sample, ps-account-failover-priority-update.md to new script in /common/ps-account-failover-priority-update.ps1
- Change existing sample, ps-account-update.md to point to new script in /common/ps-account-update.ps1

Delete the following articles from /scripts/powershell/
- create-and-configure-cassandra-database.md
- create-graph-database-account-powershell.md
- create-mongodb-database-account-powershell.md
- create-cassandra-database-account-cli.md

SnehaGunda commented:
> Hi Sneha. It seems the only remaining things was making sure the old PS scripts for Cassandra (2), Gremlin, MongoDB were deleted from docs repo. I have this on my To Do so I think we are ready to merge.
> 
> Please take one more pass at the unresolved comments and go ahead and do sign-off. Once this is merged. I will do the following on the docs side.
> 
> Create/update the following articles:
> 
> * New Samples page to point to /common/ps-account-firewall-create.ps1
> * **New Samples page to point to /common/ps-account-firewall-create.ps1** - This is same script as the above line did you had a different name? 
> * Change existing sample, ps-account-failover-priority-update.md to new script in /common/ps-account-failover-priority-update.ps1
> * Change existing sample, ps-account-update.md to point to new script in /common/ps-account-update.ps1
> 
> Delete the following articles from /scripts/powershell/
> 
> * create-and-configure-cassandra-database.md
> * create-graph-database-account-powershell.md
> * create-mongodb-database-account-powershell.md
> * **create-table-database-account-powershell.md** - Please delete this file too as discussed above. 
> * **create-cassandra-database-account-cli.md** - This is CLI script, why do you have CLI in PS I didn't validate the CLI scripts.

SnehaGunda commented:
#sign-off

@sptramer can you please merge this PR.

markjbrown commented:
Hey @sptramer can you merge this PR?

Thanks.

sptramer commented on cosmosdb/sql/ps-container-create-index-none.ps1:
```diff
@@ -19,6 +58,6 @@ $ContainerProperties = @{
-    -ApiVersion "2015-04-08" -ResourceGroupName $resourceGroupName `
+New-AzResource -ResourceType $containerResourceType `
+    -ApiVersion $apiVersion -ResourceGroupName $resourceGroupName `
     -Name $resourceName -PropertyObject $ContainerProperties
```
`$resourceName` is used here, but never defined in the script.

markjbrown replied:
Fixed in next commit.

sptramer commented on cosmosdb/sql/ps-account-firewall-create.ps1:
```diff
@@ -1,6 +1,6 @@
 # Create an Azure Cosmos Account for SQL (Core) API with IP Firewall
```
Unable to get this script to run. Errored out with:

```
New-AzResource : {"code":"BadRequest","message":"Unable to parse request payload. \r\nActivityId: f0cd3e0a-bda3-47de-acf3-e009c3f8ba8e, Microsoft.Azure.Documents.Common/2.5.1"}
At C:\Users\sttramer\src\azure-docs-powershell-samples\cosmosdb\sql\ps-account-firewall-create.ps1:26 char:1
+ New-AzResource -ResourceType "Microsoft.DocumentDb/databaseAccounts"  ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [New-AzResource], ErrorResponseMessageException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.NewAzureResourceCmdlet
```

markjbrown replied:
Found the issue with this error. Have corrected the script in the next commit.

sptramer commented on cosmosdb/sql/ps-sql-ru-update.ps1:
```diff
@@ -1,10 +1,14 @@
 # Update RU for an Azure Cosmos DB SQL (Core) API database or container
```
Unable to get this script to run. Errored out with:

```
Set-AzResource : {"code":"Unauthorized","message":"Unable to authenticate the request. The request requires valid user authentication.\r\nActivityId: 13b39691-2d72-4ef8-927e-585f7a4974cc, Microsoft.Azure.Documents.Common/2.5.1"}
At C:\Users\sttramer\src\azure-docs-powershell-samples\cosmosdb\sql\ps-sql-ru-update.ps1:20 char:1
+ Set-AzResource -ResourceType $databaseThroughputResourceType `
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Set-AzResource], ErrorResponseMessageException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementat
```

markjbrown replied:
Unclear why you received this error. I've retested the script and it is working ok. Made some small changes to this script but otherwise is ok.

markjbrown commented:
New commit to this PR. Please review.

#sign-off

SnehaGunda commented:
@sptramer can you please share the status of this PR review?

sptramer commented on cosmosdb/sql/ps-container-create-index-none.ps1:
```diff
@@ -60,4 +60,4 @@ $ContainerProperties = @{
 
 New-AzResource -ResourceType $containerResourceType `
```
This command is now failing:

```
New-AzResource : {"code":"NotFound","message":"Message: {\"code\":\"NotFound\",\"message\":\"Message: {\\\"Errors\\\":[\\\"Resource Not Found\\\"]}\\r\\nActivityId: 0b9cf458-ea24-46b6-8022-8754b7c925aa
, Request URI: /apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149960s, RequestStats: \\r\\
nRequestStartTime: 2019-09-18T17:33:18.0154163Z, RequestEndTime: 2019-09-18T17:33:18.0154163Z, Number of regions attempted: 1\\r\\nResponseTime: 2019-09-18T17:33:18.0154163Z, StoreResult: StorePhysicalAddress: rntbd://10.0.0.27:17000/apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149960s, LSN: 8, GlobalCommittedLsn: 8, PartitionKeyRangeId: , IsValid: True, StatusCode: 404, SubStatusCode: 0, RequestCharge: 1, ItemLSN: -1, SessionToken: -1#8, UsingLocalLSN: False, TransportException: null, ResourceType: Collection, OperationType: Read\\r\\nResponseTime: 2019-09-18T17:33:18.0154163Z, StoreResult: StorePhysicalAddress: rntbd://10.0.0.33:19700/apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149961s, LSN: 8, GlobalCommittedLsn: 8, PartitionKeyRangeId: , IsValid: True, StatusCode: 404, SubStatusCode: 0, RequestCharge: 1, ItemLSN: -1, SessionToken: -1#8, UsingLocalLSN: False, TransportException: null, ResourceType: Collection, OperationType: Read\\r\\n, SDK: Microsoft.Azure.Documents.Common/2.5.1\"}, Request URI: /dbs/database1/colls/container1, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.5.1"}
At /home/stramer/src/azure-docs-powershell-samples/cosmosdb/sql/ps-container-create-index-none.ps1:61 char:1
+ New-AzResource -ResourceType $containerResourceType `
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : CloseError: (:) [New-AzResource], ErrorResponseMessageException
+ FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.NewAzureResourceCmdlet
```

Unsure of what's causing the issue here offhand, looks like passing the wrong name for a resource somewhere?

markjbrown replied:
fixed.

sptramer commented on cosmosdb/sql/ps-sql-ru-update.ps1:
```diff
@@ -10,7 +10,7 @@ $databaseThroughputResourceType = "Microsoft.DocumentDb/databaseAccounts/apis/da
 $containerThroughputResourceName = $accountName + "/sql/" + $databaseName + "/" + $containerName + "/throughput"
 $containerThroughputResourceType = "Microsoft.DocumentDb/databaseAccounts/apis/databases/containers/settings"
 $throughput = 500
-$updateResource = "database" # or "container"
```
Changing this to `database` results in:

```
Set-AzResource : {"code":"NotFound","message":"Entity with the specified id does not exist in the system.\r\nActivityId: 5dc35c6d-90c6-457e-b46e-f82680c06cbb, Microsoft.Azure.Documents.Common/2.5.1"}
At /home/stramer/src/azure-docs-powershell-samples/cosmosdb/sql/ps-sql-ru-update.ps1:20 char:1
+ Set-AzResource -ResourceType $databaseThroughputResourceType `
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : CloseError: (:) [Set-AzResource], ErrorResponseMessageException
+ FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.SetAzureResourceCmdlet
```

sptramer replied:
When set to `container`, it looks like this is now sending a similar error to the `cosmosdb/sql/ps-container-create-index-none.ps1` error:

```
Set-AzResource : {"code":"NotFound","message":"Message: {\"code\":\"NotFound\",\"message\":\"Message: {\\\"Errors\\\":[\\\"Resource Not Found\\\"]}\\r\\nActivityId: 172525b6-dc2e-4b31-94b4-b3c2f1d7b150, Request URI: /apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149960s, RequestStats: \\r\\nRequestStartTime: 2019-09-18T17:37:12.5450243Z, RequestEndTime: 2019-09-18T17:37:12.5550459Z, Number of regions attempted: 1\\r\\nResponseTime: 2019-09-18T17:37:12.5550459Z, StoreResult: StorePhysicalAddress: rntbd://10.0.0.27:17000/apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149960s, LSN: 8, GlobalCommittedLsn: 8, PartitionKeyRangeId: , IsValid: True, StatusCode: 404, SubStatusCode: 0, RequestCharge: 1, ItemLSN: -1, SessionToken: -1#8, UsingLocalLSN: False, TransportException: null, ResourceType: Collection, OperationType: Read\\r\\nResponseTime: 2019-09-18T17:37:12.5550459Z, StoreResult: StorePhysicalAddress: rntbd://10.0.0.33:19700/apps/24e0c5ba-5ce4-45a1-98ff-27922cebbeee/services/52eac7ed-2f4b-40af-8b72-6f5995b75421/partitions/f9534c83-87fe-4b7a-a3e0-0a6e77d4b05d/replicas/132132990133149961s, LSN: 8, GlobalCommittedLsn: 8, PartitionKeyRangeId: , IsValid: True, StatusCode: 404, SubStatusCode: 0, RequestCharge: 1, ItemLSN: -1, SessionToken: -1#8, UsingLocalLSN: False, TransportException: null, ResourceType: Collection, OperationType: Read\\r\\n, SDK: Microsoft.Azure.Documents.Common/2.5.1\"}, Request URI: /dbs/database1/colls/container1, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.5.1"}
At /home/stramer/src/azure-docs-powershell-samples/cosmosdb/sql/ps-sql-ru-update.ps1:25 char:1
+ Set-AzResource -ResourceType $containerThroughputResourceType `
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : CloseError: (:) [Set-AzResource], ErrorResponseMessageException
+ FullyQualifiedErrorId : Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.SetAzureResourceCmdlet
```

markjbrown replied:
Explained the first issue in comment below. Added note in script for users. No explanation on second issue, could be transient. Cannot repro.

SnehaGunda commented:
@markjbrown Can you please take a look at the comments?

markjbrown commented:
I have fixed the issue with ps-container-create-index-none.ps1. 

However there is no issue with ps-sql-ru-update.ps1. The first error you are seeing is if the user tries to update a resource that does not have throughput set. That is why you see the "entity with specified id does not exist error". Because the resource does not exists as a throughput resource in our RP. I have added a small comment in this script for users to explain.

I do not know what caused the second error. This could be transient. I have tested this script multiple times to confirm.

Updates coming in this next commit.

markjbrown commented:
#sign-off
